L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+
+
+## Background
+
+gRPC and protocol buffers evolved in a Google ecosystem supported by a [monorepo](https://www.wired.com/2015/09/google-2-billion-lines-codeand-one-place/) and [hermetic build system](https://arxiv.org/pdf/1702.01715.pdf). When engineers check a `.proto` file into source control using that toolchain, the resultant code generated by the protocol compiler does not end up in source control. Instead, the code is generated on demand during builds and cached in a distributed build artifact store. While such sophisticated machinery has started to become available in the open source community (e.g. [Bazel](https://bazel.build/)), it has not yet gained much traction. Instead, small Git repos and non-hermetic language-specific build tools are the norm. As a result, code generated by the protocol compiler is often checked into source control alongside the code which uses it.
+
+At the least, this results in surprises when an update to a `.proto` file does not result in an update to the behavior of the user’s code. However, when client and server code lives in separate repos, this can result in aliasing, where one repository houses generated code from an earlier version of the protocol buffer than the other.
+
+Open source users are aware of this gap in the ecosystem and are actively looking for ways to fill it. Many have settled on [protocol buffer monorepos](https://medium.com/namely-labs/how-we-build-grpc-services-at-namely-52a3ae9e7c35) as a solution to the problem, wherein all `.proto` files for an organization are placed in a single source code repository and included by all other repositories as a submodule. But even this is not a complete solution. In addition, some mechanism must be put in place for the repositories housing the client and server code to retrieve the desired protocol buffer and generate code for the target language.
+
+The protocol compiler paired with Google's build system also means that the average engineer never has to manually invoke `protoc`. Instead, when an update is made to a `.proto` file and some code file references it, the code for the protocol buffer is regenerated without any manual intervention on the part of the engineer. Compare that to today's workflow for gRPC users:
+
+ 1. Update .proto file.
+ 2. Manually regenerate code. (remembering how to use all of the CLI flags)
+ 3. Make necessary corresponding updates to code using the protocol buffer.
+ 4. Rerun.
+
+It's easy for several of those steps to slip one's mind while developing. Moreover, figuring out how to invoke `protoc` in a way that meshes with your imports can be quite difficult. Python developers in particular are unused to build-time steps such as these; it is much more common to perform them at runtime.
```

Comment:
Should we mention it's also beneficial to environments that only have access to ephemeral file systems? Like serverless.

Reply:
I'm not sure how this benefits environments with ephemeral filesystems? Anything that has the ability to host the application's main ".py" file also has the ability to host a pre-generated "_pb2.py" file. Even if running from a REPL, you'll still need a filesystem that's at least permanent enough to host the ".proto" file in order for `grpc.protos("foo.proto")` to work.

Reply:
I see. This is dangerously getting close to the "arbitrary protos" use case that I explicitly claim no support for in this document because it presents problems when new syntax is introduced to the protobuf specification. It also looks like IPython allows [inclusion of arbitrary resource files](https://stackoverflow.com/a/10628360/10598399) (like images). In that case, you could include a ".proto" file in the notebook and use `grpc.protos` to import it, but there isn't anything specific to the IPython environment there.

As for serverless, I don't actually know of any product that is so restrictive that it actually limits you to a single textbox. [Not appengine](https://cloud.google.com/appengine/docs/flexible/python/configuration-files). [Not Cloud Functions](https://github.com/GoogleCloudPlatform/python-docs-samples/tree/master/functions/helloworld). [Not Cloud Run](https://github.com/grpc-ecosystem/grpc-cloud-run-example/tree/master/python). [Not AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/code-editor.html). It's not a usage that I want to support or advertise.

Reply:
Sorry, I wasn't express it clear enough. A single textarea is one setup option, but it is not the only way. At least AWS Lambda allows user to supply code from its UI. I'm trying to think more scenarios that benefited by this work.